substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,231 @@
+    public String[] retrievePremainArgs(String[] args) {
+        List<String> mainArgs = new ArrayList<>();
+
+        for (String arg : args) {
+            if (arg.startsWith(PREMAIN_OPTION_PREFIX)) {
+                String premainOptionKeyValue = arg.substring(PREMAIN_OPTION_PREFIX.length());
+                String[] pair = premainOptionKeyValue.split(":");
+                if (pair.length == 2) {
+                    premainOptions.put(pair[0], pair[1]);
+                }
+            } else {
+                mainArgs.add(arg);
+            }
+        }
+        return mainArgs.toArray(new String[0]);
+    }
+
+    public void invokePremain() {
+        for (PremainMethod premainMethod : premainMethods) {
+            try {
+                Object[] args = premainMethod.args;
+                if (premainOptions.containsKey(premainMethod.className)) {
+                    args[0] = premainOptions.get(premainMethod.className);
+                }
+                // premain method must be static
+                premainMethod.method.invoke(null, args);
+            } catch (Throwable t) {
+                VMError.shouldNotReachHere("Fail to execute " + premainMethod.className + ".premain", t);
+            }
+        }
+    }
+
+    /**
+     * This class is a dummy implementation of {@link Instrumentation} interface. It serves as the
+     * registered premain method's second parameter. At native image runtime, no actual
+     * instrumentation work can do. So all the methods here are empty.
+     */
+    public static class NativeImageNoOpRuntimeInstrumentation implements Instrumentation {
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer, boolean canRetransform) {
+        }
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer) {
+        }
+
+        @Override
+        public boolean removeTransformer(ClassFileTransformer transformer) {
+            return false;
+        }
+
+        @Override
+        public boolean isRetransformClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void retransformClasses(Class<?>... classes) throws UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isRedefineClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void redefineClasses(ClassDefinition... definitions) throws ClassNotFoundException, UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isModifiableClass(Class<?> theClass) {
+            return false;
+        }
+
+        @Override
+        public Class<?>[] getAllLoadedClasses() {
+            return new Class<?>[0];
+        }
+
+        @Override
+        public Class<?>[] getInitiatedClasses(ClassLoader loader) {
+            return new Class<?>[0];
+        }
+
+        @Override
+        public long getObjectSize(Object objectToSize) {
+            return 0;
```

Comment:
We should either return the actual value by looking into Native Image internals, or something completely unusable such as `Long.MIN_VALUE`. Is this information used to transform classes?

Reply:
This method is rarely used. 
It is not used by class transformation.
I've searched the source of [opentelemetery java instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and didn't find any usage as well.
In OpenJDK, it returns `oop->size() * wordSize`. What is the similar function in substratevm? We can do the similar calculation so it will be perfectly compatible with JVM. But considering the rarely usage, `Long.MIN_VALUE` may be also acceptable for now.

Reply:
I'd suggest `0` in case anyone is summing over these values, otherwise I think they'll get negative overflow(?) by using MIN_VALUE

Reply:
I set it to -1 as it's both invalid and won't get overflowed easily.